convert.go

```diff
@@ -4,112 +4,198 @@ import (
+	if !ok {
+		return nil, errIncorrectNetAddr
+	}
+	return FromIP(ac.IP)
+}
+
+func getAddrParser(net string) (AddrParser, error) {
+	addrParsersLock.Lock()
+	defer addrParsersLock.Unlock()
+
+	parser, ok := addrParsers[net]
+	if !ok {
+		return nil, fmt.Errorf("unknown network %v", net)
+	}
+	return parser, nil
+}
+
 // FromNetAddr converts a net.Addr type to a Multiaddr.
 func FromNetAddr(a net.Addr) (ma.Multiaddr, error) {
 	if a == nil {
 		return nil, fmt.Errorf("nil multiaddr")
 	}
+	p, err := getAddrParser(a.Network())
+	if err != nil {
+		return nil, err
+	}
 
-	switch a.Network() {
-	case "tcp", "tcp4", "tcp6":
-		ac, ok := a.(*net.TCPAddr)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get IP Addr
-		ipm, err := FromIP(ac.IP)
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get TCP Addr
-		tcpm, err := ma.NewMultiaddr(fmt.Sprintf("/tcp/%d", ac.Port))
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Encapsulate
-		return ipm.Encapsulate(tcpm), nil
-
-	case "udp", "upd4", "udp6":
-		ac, ok := a.(*net.UDPAddr)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get IP Addr
-		ipm, err := FromIP(ac.IP)
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get UDP Addr
-		udpm, err := ma.NewMultiaddr(fmt.Sprintf("/udp/%d", ac.Port))
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Encapsulate
-		return ipm.Encapsulate(udpm), nil
-
-	case "utp", "utp4", "utp6":
-		acc, ok := a.(*utp.Addr)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get UDP Addr
-		ac, ok := acc.Child().(*net.UDPAddr)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get IP Addr
-		ipm, err := FromIP(ac.IP)
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Get UDP Addr
-		utpm, err := ma.NewMultiaddr(fmt.Sprintf("/udp/%d/utp", ac.Port))
-		if err != nil {
-			return nil, errIncorrectNetAddr
-		}
-
-		// Encapsulate
-		return ipm.Encapsulate(utpm), nil
-
-	case "ip", "ip4", "ip6":
-		ac, ok := a.(*net.IPAddr)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-		return FromIP(ac.IP)
-
-	case "ip+net":
-		ac, ok := a.(*net.IPNet)
-		if !ok {
-			return nil, errIncorrectNetAddr
-		}
-		return FromIP(ac.IP)
+	return p(a)
+}
 
-	default:
-		return nil, fmt.Errorf("unknown network %v", a.Network())
+func getMaddrParser(name string) (MaddrParser, error) {
+	addrParsersLock.Lock()
+	defer addrParsersLock.Unlock()
+	p, ok := maddrParsers[name]
+	if !ok {
+		return nil, fmt.Errorf("network not supported: %s", name)
 	}
+
+	return p, nil
 }
 
 // ToNetAddr converts a Multiaddr to a net.Addr
 // Must be ThinWaist. acceptable protocol stacks are:
 // /ip{4,6}/{tcp, udp}
 func ToNetAddr(maddr ma.Multiaddr) (net.Addr, error) {
+	protos := maddr.Protocols()
+	final := protos[len(protos)-1]
+
+	p, err := getMaddrParser(final.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return p(maddr)
+}
+
+func registerDefaultMaddrParsers() {
```

Comment:
If we're keeping this (i.e. not using the public API for registering), could we move it up to live with the `registerDefaultAddrParsers`?